From 6129c75531b2f99395367c5e8725b87841e2172b Mon Sep 17 00:00:00 2001
From: Michael Olbrich <m.olbrich@pengutronix.de>
Date: Fri, 31 Aug 2012 14:46:47 +0200
Subject: [PATCH 21/40] v4l2object: add more properties

* JPEG restart interval
* H264 I-Frame QP Value
* H264 P-Frame QP Value
* Video Bitrate

Signed-off-by: Michael Olbrich <m.olbrich@pengutronix.de>
---
 sys/v4l2/gstv4l2object.c | 66 ++++++++++++++++++++++++++++++++++++++++++++++++
 sys/v4l2/gstv4l2object.h |  4 +++
 2 files changed, 70 insertions(+)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 2723ef5..41832b4 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -489,6 +489,46 @@ gst_v4l2_object_install_properties_helper (GObjectClass * gobject_class,
           G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS | GST_PARAM_CONTROLLABLE));
 
   /**
+   * GstV4l2Src:jpeg-restart-interval
+   *
+   * JPEG restart interval
+   */
+  g_object_class_install_property (gobject_class, PROP_JPEG_RESTART_INTERVAL,
+      g_param_spec_int ("jpeg-restart-interval", "JPEG restart interval",
+          "JPEG restart interval", 0, G_MAXINT, 0,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS | GST_PARAM_CONTROLLABLE));
+
+  /**
+   * GstV4l2Src:h264-i-frame-qp
+   *
+   * H264 I-Frame QP Value
+   */
+  g_object_class_install_property (gobject_class, PROP_H264_I_FRAME_QP,
+      g_param_spec_int ("h264-i-frame-qp", "H264 I-Frame QP",
+          "H264 I-Frame QP Value", 1, 51, 30,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS | GST_PARAM_CONTROLLABLE));
+
+  /**
+   * GstV4l2Src:h264-p-frame-qp
+   *
+   * H264 P-Frame QP Value
+   */
+  g_object_class_install_property (gobject_class, PROP_H264_P_FRAME_QP,
+      g_param_spec_int ("h264-p-frame-qp", "H264 P-Frame QP",
+          "H264 P-Frame QP Value", 1, 51, 30,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS | GST_PARAM_CONTROLLABLE));
+
+  /**
+   * GstV4l2Src:bitrate
+   *
+   * Video Bitrate
+   */
+  g_object_class_install_property (gobject_class, PROP_BITRATE,
+      g_param_spec_int ("bitrate", "Bitrate",
+          "Video Bitrate", 0, G_MAXINT, 0,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS | GST_PARAM_CONTROLLABLE));
+
+  /**
    * GstV4l2Src:norm
    *
    * TV norm
@@ -615,6 +655,24 @@ gst_v4l2_object_prop_to_cid (guint prop_id)
       cid = V4L2_CID_JPEG_COMPRESSION_QUALITY;
       break;
 #endif
+#ifdef V4L2_CID_JPEG_RESTART_INTERVAL
+    case PROP_JPEG_RESTART_INTERVAL:
+      cid = V4L2_CID_JPEG_RESTART_INTERVAL;
+      break;
+#endif
+#ifdef V4L2_CID_MPEG_VIDEO_H264_I_FRAME_QP
+    case PROP_H264_I_FRAME_QP:
+      cid = V4L2_CID_MPEG_VIDEO_H264_I_FRAME_QP;
+      break;
+#endif
+#ifdef V4L2_CID_MPEG_VIDEO_H264_P_FRAME_QP
+    case PROP_H264_P_FRAME_QP:
+      cid = V4L2_CID_MPEG_VIDEO_H264_P_FRAME_QP;
+      break;
+#endif
+    case PROP_BITRATE:
+      cid = V4L2_CID_MPEG_VIDEO_BITRATE;
+      break;
     default:
       GST_WARNING ("unmapped property id: %d", prop_id);
   }
@@ -642,6 +700,10 @@ gst_v4l2_object_set_property_helper (GstV4l2Object * v4l2object,
     case PROP_HUE:
     case PROP_GOP_SIZE:
     case PROP_JPEG_QUALITY:
+    case PROP_JPEG_RESTART_INTERVAL:
+    case PROP_H264_I_FRAME_QP:
+    case PROP_H264_P_FRAME_QP:
+    case PROP_BITRATE:
     {
       gint cid = gst_v4l2_object_prop_to_cid (prop_id);
 
@@ -757,6 +819,10 @@ gst_v4l2_object_get_property_helper (GstV4l2Object * v4l2object,
     case PROP_HUE:
     case PROP_GOP_SIZE:
     case PROP_JPEG_QUALITY:
+    case PROP_JPEG_RESTART_INTERVAL:
+    case PROP_H264_I_FRAME_QP:
+    case PROP_H264_P_FRAME_QP:
+    case PROP_BITRATE:
     {
       gint cid = gst_v4l2_object_prop_to_cid (prop_id);
 
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index d06a16c..20ecc73 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -176,6 +176,10 @@ GType gst_v4l2_object_get_type (void);
     PROP_HUE,                           \
     PROP_GOP_SIZE,                      \
     PROP_JPEG_QUALITY,                  \
+    PROP_JPEG_RESTART_INTERVAL,         \
+    PROP_H264_I_FRAME_QP,               \
+    PROP_H264_P_FRAME_QP,               \
+    PROP_BITRATE,                       \
     PROP_TV_NORM,                       \
     PROP_IO_MODE
 
-- 
1.8.3

